module/src/main/scala/com/gu/googleauth/actions.scala

```diff
@@ -52,13 +50,54 @@ trait UserIdentifier {
     UserIdentity.fromRequest(request).filter(_.isValid || !authConfig.enforceValidity)
 }
 
-trait Actions extends UserIdentifier {
+object AuthAction {
+  type UserIdentityRequest[A] = AuthenticatedRequest[A, UserIdentity]
+}
+
+/**
+  * This action ensures that the user is authenticated and their token is valid. Is a user is not logged in or their
+  * token has expired then they will be authenticated.
+  *
+  * The AuthenticatedRequest will always have an identity.
+  *
+  * @param authConfig
+  * @param loginTarget The target that should be redirected to in order to carry out authentication
+  */
+class AuthAction[A](val authConfig: GoogleAuthConfig, loginTarget: Call)
+  (implicit bodyParser: BodyParser[A], val executionContext: ExecutionContext)
+  extends ActionBuilder[AuthAction.UserIdentityRequest, A]
+    with ActionRefiner[Request, AuthAction.UserIdentityRequest]
+    with UserIdentifier {
+
+
+
+  override protected def refine[A](request: Request[A]): Future[Either[Result, AuthAction.UserIdentityRequest[A]]] =
+    Future.successful(
+      userIdentity(request)
+        .map(userID => new AuthenticatedRequest(userID, request))
+        .toRight(sendForAuth(request))
+    )
+
+  /**
+    * Helper method that deals with sending a client for authentication. Typically this should store the target URL and
+    * redirect to the loginTarget. There shouldn't really be any need to override this.
+    */
+  def sendForAuth[A](request: RequestHeader)(implicit ec: ExecutionContext) =
```

Comment:
A search of our repos does find some instances of this being used elsewhere, though in at least once case it's using rather an old version!